cdc/generation.cc

```diff
@@ -123,176 +123,198 @@ const std::vector<token_range_description>& topology_description::entries() cons
     return _entries;
 }
 
-static stream_id make_random_stream_id() {
+static stream_id create_stream_id(dht::token t) {
     static thread_local std::mt19937_64 rand_gen(std::random_device().operator()());
     static thread_local std::uniform_int_distribution<int64_t> rand_dist(std::numeric_limits<int64_t>::min());
 
-    return {rand_dist(rand_gen), rand_dist(rand_gen)};
+    return {dht::token::to_int64(t), rand_dist(rand_gen)};
 }
 
-/* Given:
- * 1. a set of tokens which split the token ring into token ranges (vnodes),
- * 2. information on how each token range is distributed among its owning node's shards
- * this function tries to generate a set of CDC stream identifiers such that for each
- * shard and vnode pair there exists a stream whose token falls into this
- * vnode and is owned by this shard.
- *
- * It then builds a cdc::topology_description which maps tokens to these
- * found stream identifiers, such that if token T is owned by shard S in vnode V,
- * it gets mapped to the stream identifier generated for (S, V).
- */
-// Run in seastar::async context.
-topology_description generate_topology_description(
-        const db::config& cfg,
-        const std::unordered_set<dht::token>& bootstrap_tokens,
-        const locator::token_metadata& token_metadata,
-        const gms::gossiper& gossiper) {
-    if (bootstrap_tokens.empty()) {
-        throw std::runtime_error(
-                "cdc: bootstrap tokens is empty in generate_topology_description");
-    }
-
-    auto tokens = token_metadata.sorted_tokens();
-    tokens.insert(tokens.end(), bootstrap_tokens.begin(), bootstrap_tokens.end());
-    std::sort(tokens.begin(), tokens.end());
-    tokens.erase(std::unique(tokens.begin(), tokens.end()), tokens.end());
+// Assumes t is smaller than 2^63 - 1
+static
+dht::token next_token_no_wrap_around(dht::token t) {
+    return dht::token::from_int64(dht::token::to_int64(t) + 1);
+}
 
-    std::vector<token_range_description> entries(tokens.size());
-    int spots_to_fill = 0;
+static
+dht::token smallest_token() {
+    return dht::token::from_int64(std::numeric_limits<int64_t>::min());
```

Comment:
This is actually an illegal token. The first legal token is numeric_limits::min() + 1.

Reply:
You're right. I will fix. BTW using ring_position_range_sharder would have the same problem because the sharder would return range starting with dht::minimum_token().